app/src/main/java/io/github/cmput301f19t19/legendary_fiesta/FirebaseHelper.java

```diff
@@ -60,6 +60,123 @@ public void checkUserExists(String name, final FirebaseCallback<QuerySnapshot> c
+                requests.add(doc.toObject(FriendRequest.class));
+            }
+            callback.onSuccess(requests);
+        }).addOnFailureListener(callback::onFailure);
+    }
+
+    /**
+     * add new friend request to the database
+     *
+     * @param fromUID
+     * @param toUID
+     * @param callback
+     */
+    public void sendFriendRequest(String fromUID, String toUID, final FirebaseCallback<Void> callback) {
+        FriendRequest friendRequest = new FriendRequest(new Date(), fromUID, false, toUID);
+        db.collection("requests").document().set(friendRequest)
+                .addOnSuccessListener(callback::onSuccess)
+                .addOnFailureListener(callback::onFailure);
+    }
+
+    /**
+     * flip the friend request (either approve or deny)
+     *
+     * @param fromUID
+     * @param toUser
+     * @param approve
+     */
+    public void flipFriendRequest(String fromUID, User toUser, boolean approve, final FirebaseCallback<Void> callback) {
+        db.collection("requests").whereEqualTo("from", fromUID).whereEqualTo("to", toUser.getUid())
+                .get()
+                .addOnSuccessListener(queryDocumentSnapshots -> {
+                    List<DocumentSnapshot> documentSnapshots = queryDocumentSnapshots.getDocuments();
+                    DocumentSnapshot request;
+                    // there should be only one request for each fromUID <-> toUID
+                    if (documentSnapshots.size() != 1) {
+                        callback.onFailure(new Exception("Unexpected number of requests: "));
+                        return;
+                    }
+                    request = documentSnapshots.get(0);
+                    // request shouldn't be null
+                    assert request != null;
+                    if (!approve) {
+                        // if it's reject, we then remove the request
+                        db.collection("requests").document(request.getId()).delete()
+                                .addOnFailureListener(callback::onFailure)
+                                .addOnSuccessListener(callback::onSuccess);
+                        return;
+                    }
+                    // approve: update the status to true
+                    db.collection("requests").document(request.getId()).update("status", true)
+                        .addOnSuccessListener(aVoid -> {
+                            toUser.acceptFollowRequest(fromUID);
+                            db.collection("users").document(toUser.getUid())
+                                    .update("followedBy", toUser.getFollowedBy())
+                                    .addOnSuccessListener(callback::onSuccess)
+                                    .addOnFailureListener(callback::onFailure);
+                        })
+                        .addOnFailureListener(callback::onFailure);
+                })
+                .addOnFailureListener(callback::onFailure);
+    }
+
+    /**
```

Comment:
Please write a comment for what this does

Reply:
Addressed